File: the-animals/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement {
    private MapDirection orientation;
    private Vector2d position;

    public Animal() {
        this(new Vector2d(2, 2));
    }
    public Animal(Vector2d initialPosition) {
        this.orientation = MapDirection.NORTH;
        this.position = initialPosition;
    }
    public Vector2d getPosition() {
        return this.position;
    }
    public MapDirection getOrientation() {
        return this.orientation;
    }
    public String toString() {
        return this.orientation.toString();
    }

    @Override
    public String getResourceName() {
        return switch(this.orientation) {
            case NORTH -> "up.png";
            case EAST -> "right.png";
            case SOUTH -> "down.png";
            case WEST -> "left.png";
        };
    }

    public boolean isAt(Vector2d position) {
        return this.position.equals(position);
    }
    void move(MoveDirection direction, MoveValidator validator) {
        Vector2d newPosition = switch (direction) {
            case FORWARD -> this.position.add(this.orientation.toUnitVector());
            case BACKWARD -> this.position.subtract(this.orientation.toUnitVector());
            case RIGHT -> {
                this.orientation = this.orientation.next();
                yield this.position;
            }
            case LEFT -> {
                this.orientation = this.orientation.previous();
                yield this.position;
            }
        };
        if(validator.canMoveTo(newPosition)) {
            this.position = newPosition;
        }
    }
}

File: lab_bonus2/src/main/kotlin/Animal.kt
class Animal(
    var position: Vector2d = Vector2d(2, 2),
    var orientation: MapDirection = MapDirection.NORTH
) {
    fun isAt(position: Vector2d): Boolean = this.position == position

    fun move(direction: MoveDirection) {
        val newPosition = when (direction) {
            MoveDirection.FORWARD -> position + orientation.toUnitVector()
            MoveDirection.BACKWARD -> position - orientation.toUnitVector()
            MoveDirection.RIGHT -> {
                orientation = orientation.next()
                position
            }

            MoveDirection.LEFT -> {
                orientation = orientation.previous()
                position
            }
        }
    }

    override fun toString(): String = orientation.toString()
}